LBF forms client side validation

sharonco-matrix opened the pull request:
The following branch was created to solve the client side validation on the LBF forms using the validate.js library (https://validatejs.org/).
Changes:
- Added field in Administration->Layouts->"?" : to choose from a predefined rules , the one that will be applied for this specific field

- Added a pre-filled list (LBF_Validation) with some basic/examples on validation rules 

- More rules can be added using the syntax of validate.js library (constraints)

- We added some custom validation rules (coded inside the validate_extend.js)

- Added a new global variable "new_validation" to keep backwards compatibility. If we want to use the new validation mechanism the variable should be set to true

- Refactor the demographics (new/edit) and history to use this new mechanism

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,134 @@
+
+            ?>
+            var valid = true ;
+            var constraints = <?=$constraints?>;
```
Don't use PHP short tags as they might be disabled.  Use <?php echo ... ?> instead.

sharonco-matrix replied:
Fixed!

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,134 @@
+                 }
+
+                //bind hide function on focus/select again
+                $(input).on('click focus select', function(){
```
I think .on() requires a recent version of jquery.  Are you SURE that is always the case here?

sharonco-matrix replied:
The files that are using now the new validation (demographics, new patient and history are using:
<script type="text/javascript" src="../../../library/js/jquery-1.9.1.min.js"></script>

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,134 @@
+
+            <?php
+
+            $constraints = LBF_Validation::generate_validate_constraints($form_id);
```
Add some comments here to explain what's going on.  Many developers will have no experience with this validation library and will not enjoy maintaining your code unless you help them.

sharonco-matrix replied:
Done!

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,134 @@
+
+<script language='JavaScript'>
+
+    function submitme(new_validate,e,form_id) {
```
Don't make this hard on other developers. Add comments to document what the function does. Be sure to note that it calls the existing validate() function.

sharonco-matrix replied:
Done!

sunsetsystems commented on library/validation/LBF_Validation.php:
```diff
@@ -0,0 +1,49 @@
+        $fres = sqlStatement(
+             "SELECT layout_options.*,list_options.notes as validation_json 
+              FROM layout_options  
+              LEFT JOIN list_options ON layout_options.validation=list_options.option_id
```
This doesn't look right. Surely it matters what list_options.list_id is?

sharonco-matrix replied:
We need to  the layout_options to the list_options. in order to get the "rule" that is inside the notes of the list table.
list_options.list_id this is the foreign key we are inserting in the lbf layouts.

So i'm not sure what you meant in your question, can you explain it please?

sunsetsystems replied:
What I mean is in the ON clause you have to specify the list_id of the list that you are matching on.  Otherwise a different list might have an option_id that matches.

sharonco-matrix replied:
Fixed :)
Thanks for that :)

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,134 @@
+            var constraints = <?=$constraints?>;
+            var error_msg ='<?=xl('is not valid')?>';
+            var form = document.querySelector("form#"+form_id);
+            var elements = validate.collectFormValues(form);
```
This is confusing.  Looks like there is a validate object as well as the caller's existing validate() function.  It's not obvious what's being referenced here.

sharonco-matrix replied:
Added comments

sunsetsystems commented:
Not done yet... will probably continue over the weekend.

sharonco-matrix commented:
@sunsetsystems :+1: 
Thank you very much for reviewing the code and for your comments.
I added descriptions and fixed some things
Please let me know if there is more to do
Have a great day!

sunsetsystems commented:
Thanks for the updates.  I still want to go over it some more and will do that soon.

sunsetsystems commented on library/options.inc.php:
```diff
@@ -207,6 +207,13 @@ function generate_select_list($tag_name, $list_id, $currvalue, $title, $empty_na
+    // add default selected to 'Unassigned' for multiple select
+    if(!$got_selected && !$got_selected_backup && $multiple) {
+
+        $s = str_replace("value=''", "value='' selected ", $s);
```
Not an expected practice to scan/replace the output. Better to mark as selected in the code above.

sunsetsystems commented on library/globals.inc.php:
```diff
@@ -1372,6 +1372,12 @@
       '',                               // default
       xl('Email address, if any, to receive emergency login user activation messages.')
     ),
+     'new_validate' => array(
```
Others have not been doing this, but I prefer adding "gbl_" in front of these names to avoid name collisions .  So 'gbl_new_validate'.

olegsv replied:
@sunsetsystems is this really necessary? We have added several global configuration parameters into codebase , and there was no such requirements. It's a bit strange when two code reviewers have different naming rules for parameters.

sunsetsystems replied:
It's only a suggestion, you can ignore it.  I'm sure you understand the point.  Thanks.

sunsetsystems commented on library/js/fancybox/jquery.fancybox-1.2.6.js:
```diff
@@ -33,7 +33,7 @@
 
 	var elem, opts, busy = false, imagePreloader = new Image, loadingTimer, loadingFrame = 1, imageRegExp = /\.(jpg|gif|png|bmp|jpeg)(.*)?$/i;
-	var ieQuirks = null, IE6 = $.browser.msie && $.browser.version.substr(0,1) == 6 && !window.XMLHttpRequest, oldIE = IE6 || ($.browser.msie && $.browser.version.substr(0,1) == 7);
+	var ieQuirks = null, IE6 = false, oldIE = false;
```
What's this for?  Does FancyBox need upgrading?

sharonco-matrix replied:
Yes Fancy box needs upgrade :)

This lines are for IE6

olegsv replied:
1.2.6 was the last  version of Fancybox before rewriting to 1.3.x and the 1.x branch was abandoned in 2010.
So yes, we need to upgrade it sometime.

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,145 @@
+
+    //Use the old validation script if no parameter sent (backward compatibility)
+    //if we want to use the "old" validate function (set in globals) the validate function that will be called is the one that
+    // was on code up to today (the validat library was not loaded ( look at the top of this file)
```
Thanks for the explanation.

sharonco-matrix replied:
No, because the validation library is not loaded, it loads only if the "new validation is used"

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,145 @@
+            var form = document.querySelector("form#"+form_id);
+            //gets all the "elements" in the form and sends them to the validate library
+            //for more information @see https://validatejs.org/
+            var elements = validate.collectFormValues(form);
```
Possible name collision on "validate"?

sharonco-matrix replied:
validate-> from the new validation is inside the validate.js library and its not loaded unless the global to use this validation is set to on

sunsetsystems replied:
Yes but if it is set to on, what ensures that the old validate function is not also there?

sharonco-matrix replied:
You are right, we dont see an error because it take the last one in code.
What do you suggest? We cannot change the name of the validate function in the library. So we could only change the name of the "old" validation .

sunsetsystems replied:
OK doing a bit of research, it seems that functions, like variables, can re-use the same name and the last-defined function is the one that is used.  So I would leave it as-is but add a comment to this effect to avoid confusion.

sunsetsystems commented on sql/4_2_2-to-4_3_1_upgrade.sql:
```diff
@@ -181,3 +181,21 @@ CREATE TABLE `valueset` (
+#IfMissingColumn layout_options validation
+ALTER TABLE layout_options ADD COLUMN validation varchar(100) default NULL;
+#EndIf
+#IfNotRow list_options option_id LBF_Validations
```
If the user deleted all these validations then this will add them back.  Better to use:
IfNotRow2D list_options list_id lists option_id LBF_Validations

sharonco-matrix replied:
I don't understand what the alter table to add the column validations in layout options has to do with the list_options table data?
We need that column in the DB structure even if its blank .. what i'm missing :)

sunsetsystems replied:
My comment is not about the ALTER TABLE.  It is to suggest changing this line:
#IfNotRow list_options option_id LBF_Validations

sunsetsystems commented on interface/patient_file/summary/demographics_full.php:
```diff
@@ -744,6 +737,12 @@ function end_group() {
 <?php include $GLOBALS['fileroot']."/library/options_listadd.inc"; ?>
 
+<?php /*Include the validation script and rules for this form*/
+$form_id="DEM";
```
What matches this?  I don't see a DEM in the validation stuff.

sharonco-matrix replied:
This is used to bring the constraint for the fields in each form. 
We did some refactor of the submit me function and we took it out to another file that we are including. we are also using the same  function to do the query to bring the fields and the parameter is the form id so we are just defining it

sunsetsystems replied:
I guess this will be more clear when I test. Do you have any constraints defined for the DEM layout?

olegsv replied:
No we don't yet.

sunsetsystems commented:
OK after you have addressed comments so far I will try testing. Do you have any tips for doing that?

sharonco-matrix commented:
My tips:
Phase 1:
+Change the global (security->new_validation->checkit
+Add validation for some fields in demographics or history LBF forms 
       ->Inside the "?" window
+ Start using only the validation rules that we added in the list (LBF_Validations lists)
+Play with it and enjoy the results
Phase 2:
+Try to add more rules following the validation https://validatejs.org/ usage 
    -lists->LBF_Validation
+Add this rules to demographics or history forms

****++++Be patient if you find some bug :)
++++ENJOY THE RIDE ++++++****

sunsetsystems commented on sql/4_2_2-to-4_3_1_upgrade.sql:
```diff
@@ -181,3 +181,22 @@ CREATE TABLE `valueset` (
+ALTER TABLE layout_options ADD COLUMN validation varchar(100) default NULL;
+#EndIf
+
+#IfNotRow2D list_options option_id LBF_Validations
```
#IfNotRow2D requires 5 arguments.  Please check my suggestion again.

matrix-amiel replied:
I don't understand the comment
We used #IfNotRow. Why we should use #IfNotRow2D ? 
thanks

matrix-amiel replied:
sorry my mistake , I fixed the code according your suggestion.

sunsetsystems commented on interface/super/edit_layout.php:
```diff
@@ -47,6 +47,13 @@
 }
 
+// Include predefined Validation Rules from list
+$validations=[];
```
This does not work with PHP 5.3.  Do this instead:
$validations = array();

sunsetsystems commented:
Tried testing but didn't get very far.  This was with PHP 5.3 in Ubuntu 12.04 which is still in support.

First was the [] array literal syntax mentioned above.  I corrected that and then selected validation rules for first/middle/last name and email addresses.  Tried to create a patient  like this:

![matrix_2](https://cloud.githubusercontent.com/assets/386879/16606586/085a1312-42ee-11e6-8c70-6a91fd4e12a4.png)

but on clicking Create New Patient nothing happened.  The browser reported this JavaScript error:

![matrix_3](https://cloud.githubusercontent.com/assets/386879/16606617/4846d1cc-42ee-11e6-83a6-a92a4d67f2d8.png)

sharonco-matrix commented:
dis you changed the global var?
Phase 1:
+Change the global (security->new_validation->checkit)

sunsetsystems commented:
Yes.

sunsetsystems commented:
The error BTW is "event not defined".  Ignore the Telemetry thing, I see that a lot.

sharonco-matrix commented:
Hi, i would like to help you debbug this, but i need to reproduce that bug, do you have skype where we can have a small chat? or can you send me the steps you did in order to install this branch so i can do exactly the same and see where its failing 
Thanks

sunsetsystems commented:
Yes I am sunsetsystems on Skype, but have to turn in for tonight.  Steps were just to download the zip file as of July 3 and do a basic install, choose some basic rules as noted above for the demographics layout, and attempt to add a patient.  If you are unable to reproduce then ping me tomorrow and we'll go from there.

sharonco-matrix commented:
SO i will tell you exactly what i did and i cannot reproduce:
1. Cloned openemr
2.Installed openemr
3.Checkout SC/LBFValidation branch
4.Run sqlupgrade
5. Changes globals->security->use new validation
5.Add validation rules to name and last name (only characters rule: names)
6. C
![onlylettersedit](https://cloud.githubusercontent.com/assets/17809889/16649373/61029558-4441-11e6-8c3a-87ff793a27db.png)
![onlylettersnew](https://cloud.githubusercontent.com/assets/17809889/16649376/647fe8ca-4441-11e6-96a2-224ba4b2c3e7.png)

reated a new patient (wrong values) -> VALIDATION WORKED AS EXPECTED
7. Fix the values -> PATIENT SAVED OK
8. Edit the patient-> (wrong values) ->-> VALIDATION WORKED AS EXPECTED
7. Fix the values -> PATIENT SAVED OK

No js errors related to this change.
Please let me know if it worked for you

sunsetsystems commented:
Try it with Firefox.  Your code is assuming a global event object and that does not work for all browsers.
http://javascript.info/tutorial/obtaining-event-object

sunsetsystems commented on interface/new/new_comprehensive.php:
```diff
@@ -773,52 +779,57 @@ function end_group() {
+
+    var check = function() {
+      <?php if($GLOBALS['new_validate']){?>
+            var valid = submitme(<?php echo $GLOBALS['new_validate'] ? 1 : 0;?>,event,"DEM");
```
Looks like this needs fixing to define event object.

sharonco-matrix commented:
I just tried in firefox and i see the validation working and i cannot see the error in the console: (just a few warnings not related to the implementation)
Lets try to  have a skype conversation. As i cannot reproduce i cannot fix it 
![openemr 2016-07-10 08-06-31](https://cloud.githubusercontent.com/assets/17809889/16711792/8c9dd7c2-4676-11e6-8377-03b6803feb1e.png)

sunsetsystems commented:
Really, Firefox worked for you?  OK sure I'll be in and out today.  "sunsetsystems" on Skype.  But first, can I get to your test site from the Internet?  It would be helpful if I can try that.

sharonco-matrix commented:
I could reproduce the error it was on new patient creation, edit worked good. I just did a commit with the fix.
Hope this will work for you also!
Have a great day!

sunsetsystems commented:
Great, I'll check it out!

sunsetsystems commented:
Seems to test well.  Two things I think should be fixed:
1. As previously noted, change "$validations=[]" to "$validations = array()" in edit_layout.php. PHP 5.3 support is not very important, but let's not break it for no reason at all.
2. When adding a validation to the layout the "?" does not immediately change to a green background as it does with the hide conditions. It does change after the form is saved, but better to be consistent.
When those are done I think we are good to go.

sharonco-matrix commented:
Both changes have been pushed :)

sunsetsystems commented on interface/patient_file/summary/demographics_full.php:
```diff
@@ -339,14 +341,7 @@ function validate(f) {
-function submitme() {
- var f = document.forms[0];
- if (validate(f)) {
-  somethingChanged = false; // turn off "are you sure you want to leave"
```
Whoops, found a problem here.  You zapped this part.  when submitting the form, somethingChanged must be set to false so the beforeunload event handler does not complain about leaving the page.

sharonco-matrix replied:
Sorry my bad.. lost this line in the merge...
Working good now :)

sunsetsystems commented:
Well I got the commit prepared and then found a bug.  See above.

sunsetsystems commented on library/validation/validation_script.js.php:
```diff
@@ -0,0 +1,174 @@
+    /*form id: used to get the validation rules*/?>
+
+    function submitme(new_validate,e,form_id) {
+        somethingChanged = false;
```
Not quite.  You only want to do this if validation is successful and the form will be submitted.  If the submit does not happen then somethingChanged must be left as-is, otherwise there will be no warning if the user then abandons the form.

Also would you put the standard comment header at the top of this file to indicate author, GPL license etc.?

sharonco-matrix replied:
Dont go anywhere 10 minutes i'm commiting :)

sunsetsystems replied:
Try library/acl.inc for a sample.

sharonco-matrix replied:
Better ? :)

sunsetsystems commented:
OK I think that will work, thanks!

sunsetsystems commented:
Looks like sourceforge is down so I will commit this tomorrow.  Thanks for the very useful contribution!
